transhumandesign/kag-base
# Mounted Bow and other vehicle changes

## Status

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.
 
## Listing (Includes all commits)

[changed] Mounted Bow cannot fall through small gaps anymore.
[changed] Mounted Bow can now collide with platforms, doors, etc. but not with larger blobs.
[changed] Mounted Bow bought from Trader now can be converted by picking it up.
[changed] Airship and Bomber now show healthbar, can be converted and don't decay if left alone.
[changed] Mounted Bow which is attached to another vehicle can now always be detached via button.
[changed] Catapult and Ballista which is attached to another vehicle can now be detached if touching ground or when airborne.
[changed] Vehicles' ammo counter is not hidden in single player anymore.
[changed] You will detach from or let go of vehicles that aren't from your team.
[fixed] Mounted Bow doesn't look bugged when facing the other direction.
[fixed] Vehicles attached to other vehicles don't look bugged when facing the other direction.
[fixed] Bow on Warboat now gets converted along with the Warboat.
[fixed] Converting a Vehicle that is attached to another vehicle will detach instead of converting the other vehicle.

**Team Checking Commit:** 

[changed] You can use Storage and Saw if it is in team 255.
[changed] You can change class at Tent from team 255.
[changed] Longboat doesn't Decay if left alone.
[changed] You can interact with vehicles if they are in team 255.
[fixed] Inconsistent team checking in Hall.as, HallCommon.as, MigrantBrain.as, SeatHop.as, SeatsGUI.as, VehicleConvert.as, RedBarrier.as.
[fixed] That Migrants converted to your team didn't receive your color.

**TODO:**

[Still unfixed...] Gunner in non-neutral team shooting from a neutral vehicle will cause damage to the vehicle.

**Fixed in a different PR:**

[fixed] Shooting from a Mounted Bow that is in a Catapult's mag will cause the Catapult to fly away. - #1311 
[fixed] Mounted Bow shows a wrong "load item" button after it was put in a Catapult's mag. - #1311 

## Description (Only initial commits shown)

- Mounted Bow doesn't fall through small gaps anymore. Since player can attach himself to it, the player could also fall through the gap with it, causing unwanted behavior or wall clipping. This was done by changing verticesXY.
- Bomber and Airship now show health bar. Instead of decaying if an enemy is near, it can be converted to the enemy now. As for the health bar, I think all big vehicles should show it.
- Mounted Bow/Catapult/Ballista no longer look bugged when they were attached to Bomber/Airship/Warship and then changing facing direction. Mounted Bow had this problem even when not attached.
- Mounted Bow can now always be detached from another vehicle. It was not possible before. The Mounted Bow on Warboat is unaffected and cannot be detached.
- Catapult and Ballista can now be detached if the other vehicle is either "touching ground" or "not touching ground but not touching water either". Therefore, Catapult and Ballista cannot be detached from a Warboat that is riding on water without ground. But it can be detached from Airship and Bomber as long as they are not in water. I think this feels the most practical. But of course feel free to say if you have a different opinion. 
- Mounted Bow doesn't fall through doors, platforms, etc.
- All vehicles' ammo counts are no longer hidden in single player.
- Mounted Bow can be picked up and thereby be converted to your team. Mounted Bow that has been there from the start cannot be picked up and will stay neutral and stationary, possible to be used by all teams.

## Steps to Test or Reproduce

- Place a small gap and !mounted_bow. Notice it doesn't fall into the gap anymore.
- !bomber and !airship. Notice they don't get destroyed if you go away. Notice they have health bar. 
!team 1 , notice they can be converted to Red Team.
- Go to Save the Princess, !tradingpost, !coins, buy a Mounted Bow. Shooting diagonally with it. Detach yourself and pick it up. Face the other direction. Alternatively, you can load it on a Catapult and ride the other direction. Notice it doesn't look bugged anymore.
- Go to Sandbox, !catapult, !bomber. The Catapult should be attached to the Bomber. Get seated in the Bomber and make it face the other direction. Notice the Catapult doesn't look bugged anymore.
- !bomber (or !airship or !warboat), !mounted_bow. Notice if you show buttons, there is a "Detach" button.
- !bomber, !catapult. Use the bomber to fly. Notice you can "Detach" the Catapult while in the air.
- Go to Save the Princess, !tradingpost, !coins, buy a Mounted Bow. !tent, change class to builder, !wood, build some doors and platforms. Throw the Mounted Bow into the doors and platforms. Notice that it collides with them now.
- Go to Save the Princess, !tradingpost, !coins, buy a Mounted Bow. Throw it away. !team 1. Notice that it will get converted to Red Team if you pick it up. If you go to Sandbox, /loadmap "Biurza_Black_Dog", you can notice that the Mounted Bow in the middle does not get converted to any team when attaching yourself to it.
- Go to any single player mode and use a catapult or a mounted bow. Notice that they show the ammo count now.
